### ISSUE TITLE
How do I use locally trained models to run the speaker diarization pipeline on a single file?

### ISSUE THREAD

**Issue by kan-cloud** [pyannote/pyannote-audio#453]
I know the way one loads a locally trained SAD,SCD or EMB model is:

pyannote.audio.features import Wrapper
finetuned_model = Wrapper("/path/to/train/{YOUR_PROTOCOL}.train/weights/0100.pt")

And the way to run the pretrained speaker diarization pipeline on a test file is:

pipeline = torch.hub.load('pyannote/pyannote-audio', 'dia_ami')
test_file = {'uri': 'filename', 'audio': '/path/to/your/filename.wav'}
diarization = pipeline(test_file)

Forgive me for failing to connect the dots but I can't seem to find a way to make my speaker diarization pipeline utilize my locally trained models.

**Comment by hbredin** [comment]
[This](https://github.com/pyannote/pyannote-audio/tree/develop/tutorials/pipelines/speaker_diarization) tutorial explains how to train and apply a pipeline. 

Or are you looking for the actual Python API to do that?
This is unfortunately not yet documented but this should get you started:

https://github.com/pyannote/pyannote-audio/blob/f814dbba193ba79c2b361a6c2c577318df2126ec/hubconf.py#L242-L244

I'd happily merge a PR updating the above tutorial with some example Python code.

**Comment by kan-cloud** [comment]
Sincerely appreciate the prompt response!

From my (limited) understanding, the [tutorial](https://github.com/pyannote/pyannote-audio/tree/develop/tutorials/pipelines/speaker_diarization) you referred me to (which I've followed word to word successfully)
requires me to define and annotate a protocol and test subset of that protocol. The tutorial culminates in the creation of an rttm file that contains predicted annotations of the audio files in your test data set.

My requirement is to run the pipeline on audio files one at a time as is necessary. I would like to avoid defining a test subset each time I accumulate a critical mass of audio files requiring speaker diarization. So I do think what I need is the actual Python API. Correct me if what I actually need is the tutorial.

Onto the lines of Python you highlighted, 

in the line below:

params_yml, *_ = pretrained_subdir.glob('*/*/params.yml')

is '*/*/params.yml' the address of the params.yml file generated after training your pipeline with the command in the tutorial shown below? 

$ pyannote-pipeline train --subset=development --forever ${EXP_DIR} AMI.SpeakerDiarization.MixHeadset

Once I have this params.yml file I am guessing there is no need for the sad_ami,scd_ami and emb_ami [folders](https://github.com/pyannote/pyannote-audio/tree/develop/tutorials/pipelines/speaker_diarization) referred to in the tutorial- the folders which contain the .npy files of their respective models. I am also guessing there wouldn't there be any need of the .pt files which contain the weights of the SAD,SCD and EMB models once we generated the params.yml file using pyannote-pipeline.

Finally, if I've understood correctly, the way to obtain annotations for a single test audio file in a Python notebook would be:

from pyannote.audio.pipeline.utils import load_pretrained_pipeline 
params_yml, *_ = pretrained_subdir.glob("/path/to/params.yml") 
pipeline = load_pretrained_pipeline(params_yml.parent)
test_file = {'uri': 'filename', 'audio': '/path/to/your/filename.wav'}
diarization = pipeline(test_file)


Thank you for your patience.

**Comment by kan-cloud** [comment]
Ok so after a bit of tinkering I realized that commands

import torch
pipeline = torch.hub.load('pyannote/pyannote-audio', 'dia_ami')

resulted in model being downloaded into a particular directory, for me this directory was:

Downloading pretrained model "sad_ami" to "C:\Users\Kanishk Chaturvedi\.pyannote\hub\models\sad_ami.zip".

The pipeline went to:

Downloading pretrained pipeline "dia_ami" to "C:\Users\Kanishk Chaturvedi\.pyannote\hub\pipelines\dia_ami.zip"

I wanted to use the pipelines models which I trained locally on my system to perform speaker diarization on a single file, hence I can replaced the 'sad_ami','emb_ami' and 'scd_ami' folders in '{user}/.pyannote/hub/models' with the speech_activity_detection, speaker_change_detection and speaker_embedding folders respectively which may be found in pyannote-audio/tutorials/models post training and validation. 

I removed all the pt files of the model present in the  'weights' subdirectory of model except the pt file of weight which was found to be the best during validaiton. The label of the best weight is the index of the best validation epoch and may be found in 'apply' subdirectory of model directory (see the application portion of the following [tutorial](https://github.com/pyannote/pyannote-audio/tree/develop/tutorials/models/speaker_embedding))

I also replaced the params.yml file in 'dia_ami/train/{protocol}/params.yml' with the params.yml I obtained via training the speaker diarization pipeline.

**Comment by ooza** [comment]
Hi @kan-cloud ,
Actually, I have the same goal as you.
 I've used a similar code as the one below to test a new wav file using my pertained pipeline but I got this error:  config.yml file not found! Should we provide this file?
How did you resolve this ?
> My requirement is to run the pipeline on audio files one at a time as is necessary. I would like to avoid defining a test subset each time I accumulate a critical mass of audio files requiring speaker diarization. So I do think what I need is the actual Python API. Correct me if what I actually need is the tutorial.
> 
> Onto the lines of Python you highlighted,
> 
> in the line below:
> 
> params_yml, __ = pretrained_subdir.glob('_/*/params.yml')
> 
> is '_/_/params.yml' the address of the params.yml file generated after training your pipeline with the command in the tutorial shown below?
> 
> $ pyannote-pipeline train --subset=development --forever ${EXP_DIR} AMI.SpeakerDiarization.MixHeadset
> 
> Once I have this params.yml file I am guessing there is no need for the sad_ami,scd_ami and emb_ami [folders](https://github.com/pyannote/pyannote-audio/tree/develop/tutorials/pipelines/speaker_diarization) referred to in the tutorial- the folders which contain the .npy files of their respective models. I am also guessing there wouldn't there be any need of the .pt files which contain the weights of the SAD,SCD and EMB models once we generated the params.yml file using pyannote-pipeline.
> 
> Finally, if I've understood correctly, the way to obtain annotations for a single test audio file in a Python notebook would be:
> 
> from pyannote.audio.pipeline.utils import load_pretrained_pipeline
> params_yml, *_ = pretrained_subdir.glob("/path/to/params.yml")
> pipeline = load_pretrained_pipeline(params_yml.parent)
> test_file = {'uri': 'filename', 'audio': '/path/to/your/filename.wav'}
> diarization = pipeline(test_file)
> 
> Thank you for your patience.

**Comment by kan-cloud** [comment]
1. Run the following:
import torch
pipeline = torch.hub.load('pyannote/pyannote-audio', 'dia_ami') 
2. Note the name of the directory into which the folders sad_ami,scd_ami, and emb_ami were downloaded. On my system it was something like {root}\{username}\.pyannote\hub\models\sad_ami
3.  Replace the downloaded sad_ami folder with the folder speech_activity_detection found pyannote-audio/tutorials/models. **This is the folder that contains your config.yml file**. 
 4. The name of best training epoch # can be found in "speech_activity_detection\train\AMI.SpeakerDiarization.MixHeadset.train\validate_detection_fscore\AMI.SpeakerDiarization.MixHeadset.development\apply". Remove all .pt files in "speech_activity_detection\train\AMI.SpeakerDiarization.MixHeadset.train\weights" **except the one** which bears the name of the best training epoch.
5.Rename speech_activity_detection to sad_ami
6. In a similar manner replace the scd_ami and emb_ami with speaker_change_detection and speaker_embedding respectively. 7. Remove the .pt files of all but the best epoch in both of these. 
8. Rename them to scd_ami and emb_ami
9. At this point it worked for me.

**Comment by ooza** [comment]
Hi @kan-cloud,
Thanks for your prompt answer.
I've tried this:
```
sad_scores = cwd+"/experiments/sad/train/./weights/0023.pt"
scd_scores = cwd+"/experiments/scd/train/./weights/0023.pt"
emb_scores = cwd+"/experiments/emb/train/./weights/0023.pt"
method = "affinity_propagation"

pipeline = SpeakerDiarization(sad_scores = sad_scores,
                                                  scd_scores = scd_scores,
                                                   embedding = emb_scores,
                                                   method=method)

pipeline.load_params(params_file)
diarization = pipeline(test_file)
```
But I'm not sure if it gives the same result as yours or not

**Comment by kan-cloud** [comment]
My way of verifying if it gives the same result or not would, to run your code till 

diarization = pipeline(test_file)

And after that running a cell with 

""""
diarization 
"""
 You'll get a visualization of the prediction.

See if the visualization is different for both cases.

**Comment by ooza** [comment]
@kan-cloud 
In case you are still interested to use fine-tuned models for testing a diarization pipeline locally, hereafter another way to do that:
```
test_file = {'uri': 'comp_010025', 'audio': cwd+'/experiments/test_files/comp_010025.wav'}

sad_scores = cwd+"/experiments/sad/train/./weights/0023.pt"
scd_scores = cwd+"/experiments/scd/train/./weights/0023.pt"
emb_scores = cwd+"/experiments/emb/train/./weights/0023.pt"
method = "affinity_propagation"

pipeline = SpeakerDiarization(
    sad_scores = {sad_scores: {'step': 0.1}},
    scd_scores = {scd_scores: {'step': 0.1}},
    embedding = {emb_scores: {'step': 0.1}},
    method=method)

pipeline.load_params(params_file)
diarization = pipeline(test_file)
```
The ``` --step``` parameters is set to 0.1 in the Speaker Diarization pipeline tutorial (to fine-tune the models), while its default value is 0.25. So to reproduce the same result (using the python code locally) as the one generated by the command ```pyannote-pipeline apply --subset=test...``` you should make sure that you are using the same overlap between consecutive windows, i.e., ```--step=0.1``` in my case.
See Hervé's comment in this [issue](https://github.com/pyannote/pyannote-audio/issues/457#issuecomment-708947868).

**Comment by kan-cloud** [comment]
@ooza Thank you so much!

**Comment by Nirannoel** [comment]
Hi
can i get to know about how to retrain pyannote audio model for our own data???
